### ISSUE TITLE
HttpFailedToOpenDatabaseConnection under load

### ISSUE THREAD

**Issue by Saghen** [CardanoSolutions/kupo#155]
### What Git revision / release tag are you using?

`v2.7.2+9bcc3d`

### Describe what the problem is?

Under load (~20 req/s on /datums, ~5 req/s on /matches), we're seeing some of our Kupo replicas experience high request latency until our health checks restart it. This seems to happen randomly with all nodes receiving the same req/s, but not the exact same requests. The logs mention `HttpFailedToOpenDatabaseConnection` while the latency rises. It seems possible that allowing for a custom DB connection limit could help to alleviate the issue. Additionally, it may be possible to improve throughput by using a queue for requesting DB connections rather than retrying.

![image](https://github.com/CardanoSolutions/kupo/assets/10467983/3bdb2527-641c-4eb7-9df1-23d7e28bcfba)

### What should be the expected behavior?

Stable latency under load

### If applicable, what are the logs from the server around the occurence of the problem?

The following logs cover just before the incident through to its peak, lining up with the graphs above: https://paste.super.fish/p/sBXTcR.txt

**Comment by KtorZ** [comment]
Hey, thanks for the thorough report (nice to see some proper logs!). 

Regarding the failed connections and `503` errors from the server; this is not a bug but a feature. The server is simply applying back-pressure down to the client. Using a queue for retrying failed request is something that will fall under the client's responsibility.  

However, I do concede that the retrying behavior is perhaps a bit exaggerated. The retries are mostly in place to cope with transient failures as it happens sometimes when opening up new SQLite connections, but the delay and number of attempts are simply too much. 

Note that while the latency should be relatively bounded too to roughly 1s. You shouldn't observe such high 20s+ delays, especially given the size of your queries (as far as I can tell, the tokens you're fetching are associated with a rather small number of TXOs). So after investigation, I noticed that there was an internal unbounded retry policy when opening new database connections. This may explain some of the high delays. 

I'll be monitoring and tuning this for the next release.

**Comment by Saghen** [comment]
Awesome, thank you! I've made an assumption that the failing connections are due to a limit being placed on the number of connections. Would it be possible to raise that limit? Also, under the same assumption, I realize that I explained what I meant here poorly:

> it may be possible to improve throughput by using a queue for requesting DB connections rather than retrying.

It seems like when Kupo fails to get a connection (`HttpFailedToOpenDatabaseConnection`) because the maximum number of connections has been reached, it will retry after a given period. This pull-based approach to utilizing the connections may lead to under-utilization since a connection could be available before the request is retried. Using a push-based approach, such as via a queue, might increase the utilization of the connections. But all of this is just an educated guess
